### PR TITLE
Specs rework

### PR DESCRIPTION
**Pull request by kkuprikov**

### CODE REVIEW

File: spec/speller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+RSpec.describe Forspell::Speller do
+  let(:custom_dict_path) {File.join(__dir__, 'fixtures', 'custom_dict.dict')}
+  let(:speller) { described_class.new('en_US', custom_dict_path) }
+  let(:custom_word) { 'somerandomword' }
+
+  subject { speller.correct?(word) }
+
+  describe 'incorrect word' do
+    let(:word) { 'notcorrect' }
+    specify {is_expected.to be_falsey}
```
Comment (zverok) [inline spec/speller_spec.rb]: Better 
```ruby
it {is_expected.to be_falsey}
```
It reads naturally: "it is expected to ...."

File: spec/sanitizer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+require 'yaml'
+require_relative '../lib/forspell/sanitizer'
+
+class TestClass
+  include Forspell::Sanitizer
+
+  def initialize
+    @custom_dictionary = []
+  end
+end
```
Comment (zverok) [inline spec/sanitizer_spec.rb]: BTW, this complicated test setup implies, that you don't need sanitizer to be includable -- it does NOT uses anything of the class it is included to. Calling it as `Sanitizer.sanitize(...)` from loader would be almost as easy, but will lessen the coupling of the code.

File: spec/reporter_spec.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  let(:word) { Forspell::Loaders::Word.new('file.rb', 5, 'typo') }
+  let(:error) { [word, ['type']] }
+
+  describe '#error' do
+    it 'should add error with suggestions' do
+      reporter.error(*error)
+      expect(reporter.instance_variable_get(:@errors).size).to eq(1)
+    end
+  end
+end
```
Comment (zverok) [inline spec/reporter_spec.rb]: This doesn't seem to test anything that reporter really does :woman_shrugging: 
I'd say the easiest way to do it is to have several contexts with "before"  that setups something (file, error and so on), and then test "what it outputs".
(The tests "when I do something, some internal variables change" are generally to be avoided, it breaks incapsulation seriously and says nothing about "how the class works")

File: spec/loaders/ruby_spec.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+  let(:path) { File.join __dir__, '..', 'fixtures', 'example_module.rb' }
+
+  it 'should load words from ruby comments in file' do
+    loader = described_class.new(file: path, text: nil)
+    words = loader.read
+    expect(words).not_to be_empty
+    expect(words.map(&:text)).to include(*%w[erraccessor erralias This in the have])
+  end
+end
```
Comment (zverok) [inline spec/loaders/ruby_spec.rb]: There are this responsibilities for loader:
1. Extract words from comments (and don't extract from code)
2. Make sure that words "coordinates" (file/line) are correct
3. Drop everything that should not be tested

(3) is probably easier to test on its own (add `BaseLoader` test, or extract again :)) to a small module, which you can just test with "this word → dropped; this word → preserved")
1 & 2 should be tested in loader's tests -- maybe with just one huge list of "it should extract those words with this coordinates".

File: spec/file_list_spec.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    subject { described_class.new(paths: paths, exclude_paths: exclude_paths) }
+
+    describe 'correct behavior' do
+      let(:paths) { ['lib'] }
+      let(:exclude_paths) { ['spec'] }
+      it 'should contain only files in lib' do
+        subject.each do |file| 
+          expect(file.start_with? 'lib').to be_truthy
+        end
+      end
```
Comment (zverok) [inline spec/file_list_spec.rb]: it is better to express as `it { is_expected.to all start_with 'lib' }` :)

File: spec/file_list_spec.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    end
+  end
+end
```
Comment (zverok) [inline spec/file_list_spec.rb]: I believe it should be several tests for edge cases like we described: if there is files or folders (and if path is `README.md Contributing.md` and exclude is `Contributing.md`), if there are some non-existent pathes, what should it do -- raise or notify or what; and what if `exclude` excludes all of the `include` :)

File: spec/reporter_spec.rb (hunk 3 above)
Comment (kkuprikov) [reply spec/reporter_spec.rb]: Whoops.
I didn't know how to include `puts` testing, but after some googling it seems to be like `STDOUT.should_receive(:puts).with(...)`.

Comment (zverok) [reply spec/reporter_spec.rb]: `expect { block }.to output('something').to_stdout` :)

File: lib/forspell/file_list.rb [hunk 7]
```diff
@@ -31,7 +31,7 @@ def each(&block)
 
     def expand_paths(path)
       if File.directory?(path)
-        Dir.glob("#{path}/**/*.{#{EXTENSION_GLOBS.join(',')}}")
+        Dir.glob(File.join("#{path}", '**', "*.{#{EXTENSION_GLOBS.join(',')}}"))
```
Comment (zverok) [inline lib/forspell/file_list.rb]:
```diff
- "#{path}"
+ path
```
:)

File: lib/forspell/word_matcher.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+      \'?                      # could start with apostrophe
+      ([a-z]|[A-Z])? # at least one letter,
+      ([[:lower:]])+   # then any number of letters,
+      ([\'\-])?        # optional dash/apostrophe,
+      ([[:lower:]])*     # another bunch of letters
+      \'?                # could end with apostrophe
+    $}x
+
+    def self.word? text
+      WORD.match(text) ? true : false
```
Comment (zverok) [inline lib/forspell/word_matcher.rb]: `? true : false` is unnecessary :)

File: lib/forspell/word_matcher.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+module Forspell
+  module WordMatcher
+    WORD = %r{^
+      \'?                      # could start with apostrophe
+      ([a-z]|[A-Z])? # at least one letter,
+      ([[:lower:]])+   # then any number of letters,
+      ([\'\-])?        # optional dash/apostrophe,
+      ([[:lower:]])*     # another bunch of letters
+      \'?                # could end with apostrophe
+    $}x
```
Comment (zverok) [inline lib/forspell/word_matcher.rb]: Align comments back to look nice

File: spec/file_list_spec.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+  describe 'non-existing paths' do
+    let(:paths) { %w[foo/foo2] }
+    let(:exclude_paths) { %w[foo/bar] }
+    subject { described_class.new(paths: paths, exclude_paths: exclude_paths).each }
+
+    it 'should raise' do 
+      expect do
+        described_class.new(paths: paths, exclude_paths: exclude_paths).each
+      end.to raise_error(Forspell::FileList::PathLoadError, 'foo/foo2')
```
Comment (zverok) [inline spec/file_list_spec.rb]: You could use `expect { subject }.to raise_error...`
Or, even better, use named subject `subject(:list) { ... }` in line 8, and here `expect { list }.to ....`

File: spec/reporter_spec.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  after do
+    FakeFS.deactivate!
+  end
+
+  describe '#parsing_error' do
+    it 'should output parsing error' do
+      reporter.parsing_error('message')
+      expect(File.read(logfile)).to eq("Parsing error in #{file}: message\n")
```
Comment (zverok) [inline spec/reporter_spec.rb]: It could be expressed this way (probably a bit more expressive):
```ruby
it { expect { ... }.to change { File.read(logfile) }.to "Parsing error in #{file}: message\n" }
```

File: lib/forspell/word_matcher.rb (hunk 8 above)
Comment (kkuprikov) [reply lib/forspell/word_matcher.rb]: Should I rewrite spec table then?
